titan-server/src/organizations/routes.rs

```diff
@@ -19,34 +20,96 @@ pub struct FindOrganizationResponse {
     users: Vec<OrganizationUser>
 }
 
-#[get("/<slug>")]
-pub fn find_organization(
-    slug: &RawStr,
-    unkso_main: UnksoMainForums,
-    unkso_titan: TitanPrimary,
-    app_config: State<config::AppConfig>
-) -> Result<Json<FindOrganizationResponse>, status::NotFound<String>> {
-    let organization_result = organizations::organizations::find_by_slug(
-        slug.as_str(),
-        unkso_titan
+#[derive(FromForm)]
+pub struct FindOrganizationByPathRequest {
+    pub group_type: String,
+    pub path: String
+}
+
+#[get("/path?<form..>")]
+pub fn find_organization_by_path(
+    form: Form<FindOrganizationByPathRequest>,
+    titan_db: TitanPrimary
+) ->  Result<Json<models::TitanOrganization>, status::NotFound<String>> {
+    let path: Vec<&str> = form.path.split("/").collect();
+
+    if path.len() == 0 {
+        return Err(status::NotFound("parameter 'path' is empty.".to_string()))
+    }
+
+    let org = organizations::organizations::find_by_path(
+        &form.group_type,
+        &path,
+        &*titan_db
+    );
+
+    if org.is_none() {
+        return Err(status::NotFound("".to_string()))
+    }
+
+    Ok(Json(org.unwrap()))
+}
+
+#[get("/<group_type>/<group>", rank = 1)]
+pub fn find_organization_group(
+    group_type: String,
+    group: String,
+    titan_db: TitanPrimary
+) -> Result<Json<models::TitanOrganization>, status::NotFound<String>> {
+    let org = organizations::organizations::find_by_path(
+        &group_type,
+        &vec![&*group],
+        &*titan_db
     );
 
-    if organization_result.is_err() {
+    if org.is_none() {
         return Err(status::NotFound("".to_string()))
     }
 
-    let organization = organization_result.unwrap();
+    Ok(Json(org.unwrap()))
+}
+
+#[get("/<group_type>/<group>/<unit>")]
```

Comment:
Look up by ID makes sense if it is known. For this use case, it's not. The front end route relies on a path of slugs to identify an organization (i.e. `/division/navy/alpha`). This offers better readability than `/organization/{id}`. The original `titan_branch` table set a unique constraint on the `slug` column. That constraint should be removed, since it is likely some organization will share the same slug:

```
/division/air-force/alpha   (air force alpha squad)
/division/navy/alpha (navy alpha squad)
```

Reply:
> What's the reason the frontend can't know the ID of the organization?

There is nothing preventing us from using `/organizations/{id}` as the front end route. It's friendlier to read `/organizations/navy` than `/organizations/45`.

> Slugs are brittle and allowing us to drop the uniqueness constraint (which, while it makes sense, isn't applicable now because there are no units in the clan with the same name, so we can safely make that assumption right now) makes the route brittle.

If there are no organizations with duplicate names, we can leave the constraint as-is. With the constraint in place, I don't see how referencing it is more brittle than using a numeric ID instead.

>  I don't think we should be nesting more than a single resource deep (i.e. /organizations/<org name>)

This nesting allows the front end use user-friendly routes (`division/navy/alpha`).  As the organization API grows, we can continue to implement new features in an idiomatic RESTful structure:

```
/organizations/{id}/users
/organizations/{id}/reports
...
```

I'm hesitant to refactor this feature, as it correctly implements the spec we previously discussed and approved. Revising it to `/organization/{id}` distracts us from meeting our development timeline without adding functional value.

Reply:
> There is nothing preventing us from using /organizations/{id} as the front end route. It's friendlier to read /organizations/navy than /organizations/45.

I agree with this. The frontend route should be a vanity route. It's okay for the frontend route to be `/organizations/navy/thing1/thing2`. The server route _should not_ match the vanity route in this case. The client app is for an end user to consume, the server is an API that a machine can consume (in a RESTful way).

> If there are no organizations with duplicate names, we can leave the constraint as-is. With the constraint in place, I don't see how referencing it is more brittle than using a numeric ID instead.

I think we should maintain the unique constraint. In the past, we've never really had units named the same across branches, so it should be a non-issue. If it becomes one, we can reevaluate that down the road.

> This nesting allows the front end use user-friendly routes (division/navy/alpha). As the organization API grows, we can continue to implement new features in an idiomatic RESTful structure:

Like I said above, nesting works in the client app, it _should not_ be in the server app.

As-is, I'm not ready to approve this PR. In fact, I think this whole endeavor has distracted us from meeting our development timeline. I'm inclined to suggest we table this refactor for the future because we don't strictly need it for a basic MVP.

Reply:
Alright, let's go ahead and remove this method in favor of the already present `/organization/<slug>` endpoint (backend only). I'll update the client to use that endpoint while keeping a friendlier url (`organizations/division/navy`).

Reply:
We'll revisit this at a later date.